cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/sql/SQLEngine.java

```diff
@@ -71,9 +79,42 @@
    * @param pullRequest the request containing information about the dataset name and schema.
    * @return {@link SQLPullDataset} instance that can be used to read records from the SQL engine.
    */
+  @Deprecated
   SQLPullDataset<StructuredRecord, KEY_IN, VALUE_IN> getPullProvider(SQLPullRequest pullRequest)
     throws SQLEngineException;
 
+  /**
+   * Creates a {@link SQLDatasetConsumer} for a given {@link SQLPushRequest} with support for
+   * a supplied{@link PushCapability}.
+   *
+   * If the return value is null, we must assume that the supplied {@link SQLPushRequest} i
+   * s not supported for this capability.
+   *
+   * @param pushRequest push request used to create a consumer
+   * @param capability the push capability we wish to use to upload
+   * @return a dataset consumer for this request using the specified capability.
+   */
+  @Nullable
+  default SQLDatasetConsumer<StructuredRecord> getConsumer(SQLPushRequest pushRequest, PushCapability capability) {
+    throw new UnsupportedOperationException("Not supported");
```

Comment:
I am wondering if default implementation can check if capability is hadoop - delegate to to getPushProvider

Reply:
Otherwise null

Reply:
Can I implement this in a follow-up PR once I tie this together with the SQL engine?

Reply:
Actually, I have a counter argument to this:

The current Hadoop implementation uses Spark methods (present in the RDDUtils class) to translate RDD->OutputFormat and InputFormat->RDD. 

If we want to make this fully "engine agnostic" I'd rather for the engine to implement this interoperability depending on the underlying engine.

Reply:
If we want to maintain the old `getPullProvider` and `getPushProvider` methods then sure, we can delegate this into the old methods.

Eventually I'd like to fully remove those old methods and migrate everything into this new API, where the engine gets to decide how to handle every capability.

Reply:
Well, I believe some people are already implementing it.
Also as soon as we keep the methods, we should use them. Otherwise it means  we still break the existing contract.

Reply:
I will remove these methods in a later PR. For now I want to get these changes in to continue moving forward with the implementation.

Reply:
Will removal work for backwards compatibility? This will break any pipelines on upgraded instance, won't it?

Reply:
SQLEngine is a plugin, we must preserve compatibility. If you really wish to change the interface, we would need to introduce a new interface and still support an old one with a wrapper.

Reply:
Good point, I had not considered existing versions of the plugin available in preview.

We can keep these methods as a fallback. so older versions of the plugin will use the old logic and as we can keep this as a fallback if the engine doesn't support capabilities (for older versions of the plugin).

Reply:
I have refactored this for backwards compatibility.

Reply:
Could you please clarify how getConsumer/getPushProvider calls would be done by platform for 0.18 engine?

Reply:
For 0.18, the implementation still calls the old methods (which just wrap InputFormatProvider/OutputFormatProvider for Hadoop operations.

I'll update the return value to `null` instead of throwing `UnsupportedOperationException`. The "fallback" is to use the legacy `getPullProvider` and `getPushProvider` methods.